apps/src/netsim/NetSimVisualization.js

```diff
@@ -0,0 +1,475 @@
+    return;
+  }
+
+  this.nodeTableChangeKey = this.shard_.nodeTable.tableChange.register(
+      this.onNodeTableChange_.bind(this));
+
+  this.wireTableChangeKey = this.shard_.wireTable.tableChange.register(
+      this.onWireTableChange_.bind(this));
+};
+
+/**
+ * Change which node we consider the 'local node' in the visualization.
+ * We go through a special creation process for this node, so that it
+ * looks and behaves differently.
+ * @param {?NetSimLocalClientNode} newLocalNode - null if disconnected
+ */
+NetSimVisualization.prototype.setLocalNode = function (newLocalNode) {
+  if (newLocalNode) {
+    if (this.localNode) {
+      this.localNode.configureFrom(newLocalNode);
+    } else {
+      this.localNode = new NetSimVizNode(newLocalNode);
+      this.entities_.push(this.localNode);
+      this.svgRoot_.find('#background_group').append(this.localNode.getRoot());
+    }
+  } else {
+    this.localNode.kill();
+  }
+  this.pullElementsToForeground();
+};
+
+/**
+ * Find a particular VizEntity in the visualization, by type and ID.
+ * @param {function} entityType - constructor of entity we're looking for
+ * @param {number} entityID - ID, with corresponds to NetSimEntity.entityID
+ * @returns {NetSimVizEntity} or null if not found
+ */
+NetSimVisualization.prototype.getEntityByID = function (entityType, entityID) {
+  return this.entities_.reduce(function (prev, cur) {
+    if (prev) {
+      return prev;
+    } else if (cur instanceof entityType && cur.id === entityID) {
+      return cur;
+    }
+    return null;
+  }, null);
+};
+
+/**
+ * Gets the set of VizWires directly attached to the given VizNode, (either
+ * on the local end or remote end)
+ * @param {NetSimVizNode} vizNode
+ * @returns {Array.<NetSimVizWire>} the attached wires
+ */
+NetSimVisualization.prototype.getWiresAttachedToNode = function (vizNode) {
+  return this.entities_.filter(function (entity) {
+    return entity instanceof NetSimVizWire &&
+        (entity.localVizNode === vizNode || entity.remoteVizNode === vizNode);
+  });
+};
+
+/**
+ * Handle notification that node table contents have changed.
+ * @param {Array.<Object>} rows - node table rows
+ * @private
+ */
+NetSimVisualization.prototype.onNodeTableChange_ = function (rows) {
+  // Convert rows to correctly-typed objects
+  var tableNodes = netsimUtils.nodesFromRows(this.shard_, rows);
+
+  // 1. Kill nodes from the visualization that are no longer in the table.
+  this.entities_.filter(function (vizEntity) {
+    return vizEntity instanceof NetSimVizNode;
+  }).filter(function (vizNode) {
+    return !tableNodes.some(function (node) {
+      return node.entityID === vizNode.id;
+    });
+  }).forEach(function (vizNode) {
+    vizNode.kill();
+  });
+
+  // 2. Add new nodes from the table into the visualization
+  tableNodes.forEach(function (node) {
+    var vizNode = this.getEntityByID(NetSimVizNode, node.entityID);
+    if (vizNode) {
+      vizNode.configureFrom(node);
+    } else {
```

Comment:
could return here, and then don't need to have the else condition inside curlies